### ISSUE TITLE
PolicyServer Host evaluates a policy before redirecting to an IdentityProvider

### ISSUE THREAD

**Issue by zinov** [PolicyServer/PolicyServer.Local#22]
I added to the policy server host application a configuration to authenticate a user vs an identity provider. In my case identity server.
If I use the the Authorize attribute with no policy specified on top of the Home Controller then I get redirected to my identity provider, but if I use Authorize["Prescribe Medication"] then it evaluates the policy with no redirect and in because the ClaimPrincipal doesn't have any claim inside. I am missing something here?
Please can you provide an example with an Identity Server integration?

**Comment by brockallen** [comment]
So you're saying if the policy fails, then the user still gets access to the action method?

**Comment by zinov** [comment]
no, basically what I am saying is because I am not getting redirected to my identity provider, then when the PolicyServerClaimsMiddleware executes the Invoke

` public async Task Invoke(HttpContext context, IPolicyServerRuntimeClient client)
        {
           //I am not getting any User here
            if (context.User.Identity.IsAuthenticated)`

So it continues with the others middlewares, and at some point, I guess MVC invokes the filters pipeline, it tries to evaluate the policy

Remember, as part of the integration, this class is injected
services.AddTransient<IAuthorizationPolicyProvider, AuthorizationPolicyProvider>();

So when it tries to resolve the policy it calls

`        public async override Task<AuthorizationPolicy> GetPolicyAsync(string policyName)
        {
            // check static policies first
            var policy = await base.GetPolicyAsync(policyName);

            if (policy == null)
            {
                policy = new AuthorizationPolicyBuilder()
                    .AddRequirements(new PermissionRequirement(policyName))
                    .Build();
            }

            return policy;
        }`

btw this signature should be like this 
**`public override async Task<AuthorizationPolicy> GetPolicyAsync(string policyName)`**

if the policy is not found as part of the mvc pipeline, then it creates the AuthorizationPolicyBuilder with the PermissionRequirement you implemented. Which by default will try to execute this method

`protected override async Task HandleRequirementAsync(AuthorizationHandlerContext context, PermissionRequirement requirement)
        {
            if (await _client.HasPermissionAsync(context.User, requirement.Name))
            {
                context.Succeed(requirement);
            }`

and there it depends on a context.User(which is not storing any claim because we were never redirected to our identity provider.

<img width="326" alt="1" src="https://user-images.githubusercontent.com/1099892/53838851-0a15ef80-3f64-11e9-9b4a-13116adc937c.png">

Maybe I am missing something, but don't know what. Please help

**Comment by zinov** [comment]
I think I found the problem. By default, my application had on the Startup the same code you have with a small difference.

Yours:

`services.AddMvc(options =>
            {
                // this sets up a default authorization policy for the application
                // in this case, authenticated users are required (besides controllers/actions that have [AllowAnonymous]
                var policy = new AuthorizationPolicyBuilder()
                    .RequireAuthenticatedUser()
                    .Build();
                options.Filters.Add(new AuthorizeFilter(policy));
            });`

Mine:

`services.AddMvc(options =>
            {
                // this sets up a default authorization policy for the application
                // in this case, authenticated users are required (besides controllers/actions that have [AllowAnonymous]
                var policy = new AuthorizationPolicyBuilder()
                    .RequireAuthenticatedUser()
                    .Build();
                options.Filters.Add(new AuthorizeFilter(policy));
            }).SetCompatibilityVersion(CompatibilityVersion.Version_2_2);`

See the SetCompatibilityVersion? [https://docs.microsoft.com/en-us/aspnet/core/mvc/compatibility-version?view=aspnetcore-2.2](Microsoft link)

Having that call in your code causes no redirection to my Identity Provider, even if had the configuration set on ConfigureServices

` services.AddAuthentication(options =>
            {
                options.DefaultScheme = CookieAuthenticationDefaults.AuthenticationScheme;
                options.DefaultChallengeScheme = "oidc";
            }
            )
            .AddCookie(CookieAuthenticationDefaults.AuthenticationScheme)
            .AddOpenIdConnect("oidc", options =>
            {
.....
.....
` 
Is this a bug?

**Comment by leastprivilege** [comment]
We will try to repro @brockallen

**Comment by leastprivilege** [comment]
I tried to repro this, but I can't...

**Comment by zinov** [comment]
I tried to reproduce it as well with the host example and it works. My code basically is a copy and paste from yours. I have this version Asp.Net Core 2.2.4.  which is not the same one you have, but I upgraded your example as well and it works. But the weird thing is that line, which I remove and it magically starts working as expected. Any idea or suggestion to point me where to look?
